Wayland: Implement cursor-shape-v1 protocol

JakobDev opened the pull request:
This PR adds implements the [Cursor shape protocol](https://wayland.app/protocols/cursor-shape-v1) on Wayland. This protocol let's the Compositor render the cursor, so the cursor looks the same on every program.

The protocol is used with the default cursor and cursors created by `glfwCreateStandardCursor`.

`tablet-unstable-v2` is needed to compile `cursor-shape-v1`.

Fixes #2668

JakobDev committed:
Wayland: Implement cursor-shape-v1 protocol

JakobDev committed:
Fix

Zamundaaa commented:
looks good to me!

matteodev8 commented:
Tested it with https://github.com/WerWolv/ImHex, the cursor size is correct now. I would say lgtm

llyyr commented on src/CMakeLists.txt:
```diff
@@ -105,6 +105,8 @@ if (GLFW_BUILD_WAYLAND)
     generate_wayland_protocol("fractional-scale-v1.xml")
     generate_wayland_protocol("xdg-activation-v1.xml")
     generate_wayland_protocol("xdg-decoration-unstable-v1.xml")
+    generate_wayland_protocol("tablet-unstable-v2.xml")
```
You could just declare a dummy `zwp_tablet_tool_v2_interface` object somewhere to avoid bringing in tablet-unstable-v2 for this, since we won't even be using it.

PrakarshPanwar commented:
Hi @JakobDev, @llyyr. I managed to make it work without `tablet-unstable-v2` by adding dummy `zwp_tablet_tool_v2_interface` struct in `wl_window.c`. How to add changes in this PR upstream?

PrakarshPanwar committed:
Added Fixes

- Removed tablet-unstable-v2 protocol
- Added dynamic theme loading
- Formatting fixes

cfnptr commented:
Hello! Any news on this? It would be awesome to have the same cursor inside the window on KDE Wayland :)

PrakarshPanwar committed:
Omit hard coded string in `registryHandleGlobal`

- Removed default in setCursorShape switch

jdkeke142 commented:
Tested this patch on Minecraft 1.21.11 + KDE Plasma 6.3.6 + NVIDIA. Cursor theme and size are now consistent between the game window and the rest of the desktop.

Applied as part of a patchset: https://github.com/jdkeke142/glfw-wayland-minecraft

PrakarshPanwar committed:
Fix cursor surface segfault bug

- Due to `loadCursorTheme` short circuit `glfwSetCursor` was causing segfault
- Fixed bug by only skipping cursor theme loading and only creating cursor surface in `loadCursorTheme`

(cherry picked from commit 10653d390de736fc44293a4646fba8d1175206aa)

PrakarshPanwar committed:
Bug fix for `glfwSetCursorWayland`

- Crash was caused due to `cursorTheme` being null
- Fix was to check for cursorTheme and adding extra if block to check for protocol support

(cherry picked from commit 9f419388f46ec1bd18b731daaa9fed5d6d57b7b7)

JakobDev committed:
Merge branch 'glfw:master' into cursorshape

JakobDev committed:
Merge pull request #1 from PrakarshPanwar/wl-cursor-shape-v1

Refinements in cursor shape protocol

JakobDev commented:
I included some changes from @PrakarshPanwar
